lib/datasource/github-releases/index.ts

```diff
@@ -66,11 +74,201 @@ export async function getReleases({
+  const releaseAsset = release.assets.find(
+    (a: GithubReleaseAsset) => a.name === releaseChecksumAssetName
+  );
+  if (!releaseAsset) {
+    return null;
+  }
+  if (digestAsset.digestedFileName) {
+    const releaseFilename = digestAsset.digestedFileName.replace(current, next);
+    const res = await http.get(releaseAsset.browser_download_url);
+    for (const line of res.body.split('\n')) {
+      const [lineDigest, lineFn] = line.split(/\s+/, 2);
+      if (lineFn === releaseFilename) {
+        return lineDigest;
+      }
+    }
+  } else {
+    const algorithm = inferHashAlg(digestAsset.currentDigest);
+    const newDigest = await getAssetDigest(releaseAsset, algorithm);
+    return newDigest;
+  }
+  logger.debug({ releaseAsset }, 'fetch');
+  return null;
+}
+
+export async function getGithubRelease(
+  apiBaseUrl: string,
+  repo: string,
+  version: string
+): Promise<GithubRelease> {
+  const url = `${apiBaseUrl}repos/${repo}/releases/tags/${version}`;
+  const res = await http.getJson<GithubRelease>(url);
+  return res.body;
+}
+
+function getDigestCacheKey(
+  { lookupName: repo, currentValue, currentDigest, registryUrl }: DigestConfig,
+  newValue: string
+): string {
+  const type = 'digest';
+  return `${registryUrl}:${repo}:${currentValue}:${currentDigest}:${newValue}:${type}`;
+}
+
+/**
+ * github.getDigest
+ *
+ * The `newValue` supplied here should be a valid tag for the GitHub release.
+ * Requires `currentValue` and `currentDigest`.
+ *
+ * There may be many assets attached to the release. This function will:
+ *  - Identify the asset pinned by `currentDigest` in the `currentValue` release
+ *     - Download small release assets, parse as checksum manifests (e.g. `SHASUMS.txt`).
+ *     - Download individual assets until `currentDigest` is encountered. This is limited to sha256 and sha512.
+ *  - Map the hashed asset to `newValue` and return the updated digest as a string
+ */
+export async function getDigest(
+  { lookupName: repo, currentValue, currentDigest, registryUrl }: DigestConfig,
+  newValue?: string
+): Promise<string | null> {
+  logger.debug(
+    { repo, currentValue, currentDigest, registryUrl, newValue },
+    'getDigest'
+  );
+  if (!currentDigest) {
+    return null;
+  }
+  if (!currentValue) {
+    return currentDigest;
+  }
+  const cacheKey = getDigestCacheKey(
+    { lookupName: repo, currentValue, currentDigest, registryUrl },
+    newValue
+  );
+  const cachedResult = await packageCache.get<string>(cacheNamespace, cacheKey);
+  // istanbul ignore if
+  if (cachedResult) {
+    return cachedResult;
+  }
+
+  const apiBaseUrl = getApiBaseUrl(getSourceUrlBase(registryUrl));
+  const currentRelease = await getGithubRelease(apiBaseUrl, repo, currentValue);
+  let digestAsset = await findDigestFile(currentRelease, currentDigest);
+  if (!digestAsset) {
+    digestAsset = await findAssetWithDigest(currentRelease, currentDigest);
```

Comment:
This might be my unfamiliarity with renovate internals, but, is it possible to provide a way to manually indicate which asset I care about, rather than searching? I nearly always know which asset I need, and letting me hint it here would 1) reduce load / speed up renovate 2) permit initial digest pinning.

I'm not sure if it's easier to add to the data model or overload one of the existing fields, but one really simplistic way to do it might be to overload `lookupName`, either in an unstructured fashion (like, `renovatebot/renovate/<assetName>`) or , to make it more structured, something like `renovatebot/renovate?assetName=<assetName>`. 

In any case, thanks for working on this, I'll definitely be using it when it lands!

Reply:
Embedding asset hints in `lookupName` is an interesting idea! It's supported by the [`regex` manager](https://github.com/renovatebot/renovate/blob/ab800f46d1d6e20030c0d1eeef0a76cc2d74ca7e/lib/manager/regex/index.ts#L16-L25), so I could use that too.

The other idea discussed in https://github.com/renovatebot/renovate/issues/7928 is extracting the asset URL as a new field (I'd considered abusing `registryUrl` for this), or extracting fields from the asset URL (e.g. `https://github.com/(?<depName>.+/.+)/releases/downloads/(?<lookupName>.+)`).
I like the ease of use of the latter: having to respecify the repository owner+name in renovate's annotation is inelegant. (URL hunting is how the [code I'm hoping to replace with renovate](https://github.com/thepwagner/action-update-dockerurl/blob/ead64ba9a168c50abcc5b3a87351513cd2ea8dc8/dockerurl/dependencies.go#L31) [works](https://github.com/thepwagner/archivist/pull/80/files), but I don't think the ability to expand variables in the URL is generalizable).

> I'm not sure if it's easier to add to the data model 

Me too (first time contributor). I'm not going to take on any of these approaches in this PR, but hopefully maintainers have some pointers or alternative ideas on the next round of review 🤞.